src/lib/math/bigint/bigint.h

```diff
@@ -645,11 +680,58 @@ class BOTAN_PUBLIC_API(2, 0) BigInt final {
        */
       void randomize(RandomNumberGenerator& rng, size_t bitsize, bool set_high_bit = true);
 
+      /**
+       * Serialize the absolute value of this BigInt as a big endian
+       * encoding.
+       *
+       * If out is smaller than the total bytes of the BigInt then
+       * an exception is thrown.
+       *
+       * If out is larger than the total bytes of the BigInt then the
+       * necessary number of zeros are prefixed to produce the desired
+       * output length
+       *
+       * Zero-padding the binary encoding is useful to ensure that other
+       * applications correctly parse the encoded value as "positive integer",
+       * as a leading 1-bit may be interpreted as a sign bit. It also is
+       * necessary when using a fixed size encoding for the integers.
+       *
+       * @param out destination byte span for the integer value
+       */
+      void serialize_to(std::span<uint8_t> out) const;
+
+      /**
+       * Serialize the value of this BigInt as a big endian encoding,
+       * always returning the specified number of bytes.
+       *
+       * Throws if the BigInt is too large to encode in the length
+       * specified.
+       */
+      template <typename T = std::vector<uint8_t>>
+      T serialize(size_t len) const {
+         // TODO this supports std::vector and secure_vector
+         // it would be nice if this also could work with std::array as in
+         //   bn.serialize_to<std::array<uint8_t, 32>>(32);
```

Comment:
Perhaps just add another overload like so:

```C++
template<size_t len>
std::array<uint8_t, len> serialize() const {
   std::array<uint8_t, len> out;
   this->serialize_to(out);
   return out;
}
```

usage would then look like this:

```C++
bn.serialize<32>();
```

... this style would be similar to the implementation in the `BufferSlicer`, where one would `bs.take<32>()`.

For call-site readability it might make sense to rename that overload to something like `serialize_as_array<len>()`.

Reply:
My take at this point is that it's probably quite rare to know at compile time the exact size of the integer being encoded. Can always revisit this issue later, and as always hard to take it back once it is in a release, so I'm going to defer.